src/files/compose/validateDappnodeCompose.ts

```diff
@@ -0,0 +1,237 @@
+import { Manifest } from "../manifest/types";
+import semver from "semver";
+import { ComposeService, Compose } from "./types";
+import { getIsCore } from "../../utils/getIsCore";
+import { params } from "./params";
+
+/**
+ * Validates against custom dappnode docker compose specs.
+ * This function must be executed after the official docker schema
+ * @param param0
+ */
+export function validateDappnodeCompose({
+  composeUnsafe,
+  manifest
+}: {
+  composeUnsafe: Compose;
+  manifest: Manifest;
+}): void {
+  const allowedNetworks = params.DOCKER_WHITELIST_NETWORKS.join(",");
+  const coreAliases = params.DOCKER_CORE_ALIASES.join(",");
+  const isCore = getIsCore(manifest);
+
+  // COMPOSE TOP LEVEL restrictions
+
+  validateComposeVersion(composeUnsafe);
+  validateComposeNetworks(composeUnsafe, allowedNetworks);
+
+  // COMPOSE SERVICE LEVEL restrictions
+
+  const cpServicesValues = Object.values(composeUnsafe.services);
+
+  validateComposeServicesKeys(composeUnsafe);
+  validateComposeServicesValues(cpServicesValues, isCore);
+  validateComposeServicesNetworks(
+    cpServicesValues,
+    allowedNetworks,
+    isCore,
+    coreAliases
+  );
+
+  // Compose service volumes: MUST BE CHECK TOGETHER WITH COMPOSE TOP LEVEL VOLUMES
+  const cpServiceVolumes = cpServicesValues
+    .map(composeService => composeService.volumes)
+    .flat();
+
+  if (cpServiceVolumes.length > 0) {
+    validateComposeAndComposeServicesVolumes(
+      composeUnsafe,
+      cpServiceVolumes,
+      isCore
+    );
+  }
+}
+
+/**
+ * Ensures the docker compose version is supported
+ */
+function validateComposeVersion(compose: Compose): void {
+  if (
+    semver.lt(
+      compose.version + ".0",
+      params.MINIMUM_COMPOSE_FILE_VERSION + ".0"
+    )
+  )
+    throw Error(
+      `Compose version ${compose.version} is not supported. Minimum version is ${params.MINIMUM_COMPOSE_FILE_VERSION}`
+    );
+}
+
+/**
+ * Ensures the docker compose networks are whitelisted
+ */
+function validateComposeNetworks(
+  compose: Compose,
+  allowedNetworks: string
+): void {
+  const cpNetworks = compose.networks;
+  if (cpNetworks) {
+    // Check there are only defined whitelisted compose networks
+    if (
+      Object.keys(cpNetworks).some(
+        networkName =>
+          params.DOCKER_WHITELIST_NETWORKS.indexOf(networkName) === -1
+      )
+    )
+      throw Error(`Only docker networks ${allowedNetworks} are allowed`);
```

Comment:
Make this simpler and log way more detailed info. Try to think as some developer reading this error, what would you wanna know?

```ts
for (const networkName of Object.keys(compose.networks)) {
  if (!allowedNetworkNames.has(networkName)) {
    throw Error(`Network ${networkName} not allowed, must only declared allowed networks ${allowedNetworks}`)
  }
}
```

Reply:
Not resolved